[SPARK-6229] Add SASL encryption to network library.

vanzin opened the pull request:
There are two main parts of this change:
- Extending the bootstrap mechanism in the network library to add a server-side
  bootstrap (which works a little bit differently than the client-side bootstrap), and
  to allow the  bootstraps to modify the underlying channel.
- Use SASL to encrypt data going through the RPC channel.

The second item requires some non-optimal code to be able to work around the
fact that the outbound path in netty is not thread-safe, and ordering is very important
when encryption is in the picture.

A lot of the changes outside the network/common library are just to adjust to the
changed API for initializing the RPC server.

Commit:
Add TransportServerBootstrap, make SASL code use it.

This commit changes the way a transport server is configured to use SASL. It
tried to be more in line with how you set up a client, by registering a
"bootstrap" that runs before the server is actually available. Due to how
servers internally work, the server bootstrap works differently from the client
one, but a similar external interface is provided to clients of the library.

The SASL backend was reorganized to use this new mechanism. Later, the
bootstrap will be modified so that it's possible to also use SASL for
encryption, which requires inserting a new channel handler into the pipeline
- the main reason why this change was necessary in the first place.

Commit:
Add SASL encryption to network library.

This change extends the "bootstrap" classes to expose the underlying netty
channel to the bootstrap implementations. This allows the SASL bootstraps
to register channel handlers that can then perform data encryption, when
that is enabled by the user.

Encryption is performed in a manner that is transparent to the other existing
channel handlers, so perhaps it's not the most efficient implementation,
especially since it requires having 2 frame decoders in the pipeline. But it's
the least intrusive way to do it.

Commit:
Add config option to enable SASL encryption.

Commit:
Make SASL encryption handler thread-safe, handle FileRegion messages.

Commit:
Consolidate ByteArrayWritableChannel, fix SASL code to match master changes.

Commit:
Make encryption thread-safe, less memory-intensive.

Two changes here:

First, wrap the code that does encryption in a FileRegion. Without resorting
to using a ThreadPerChannelEventLoopGroup, that's the only thread-safe way to
do encryption in netty, apparently, since message ordering is extremely
important here. This requires some funny accounting in the SASL FileRegion
implementation, because we can't know before hand how much data we'll have
to write.

Second, instead of encrypting the whole original message, do it in smaller
chunks, to avoid using too much memory on the sender side. So it should be
possible to transfer a large file block without having to load the whole
block into memory at once.

Commit:
Add option to only allow encrypted connections to the server.

Commit:
Fix a comment.

Commit:
Clean up some code.

vanzin commented:
I tested this with netty/encryption, netty/no encryption, and nio. Didn't really run performance tests (don't have a cluster available where I can really do that), but I don't expect things with encryption enabled to exactly fly.

SparkQA commented:
[Test build #29759 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29759/consoleFull) for   PR 5377 at commit [`cf2a605`](https://github.com/apache/spark/commit/cf2a605d02afd4f462270a0d808ff82c75ec04fd).

SparkQA commented:
[Test build #29759 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29759/consoleFull) for   PR 5377 at commit [`cf2a605`](https://github.com/apache/spark/commit/cf2a605d02afd4f462270a0d808ff82c75ec04fd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29759/
Test PASSed.

squito commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java:
```diff
@@ -62,13 +76,26 @@ public void doBootstrap(TransportClient client) {
-        saslClient.dispose();
-      } catch (RuntimeException e) {
-        logger.error("Error while disposing SASL client", e);
+      if (saslClient != null) {
```
rather than setting `saslClient = null` in the `encrypt` block, and then checking for null here, why not just use `if (!encrypt)` here?  If there is not another reason to `saslClient = null`, I think that would be clearer.

vanzin replied:
Checking for `encrypt` wouldn't work because we still want to dispose of the client instance when some error is thrown.

squito replied:
oh, sorry, I should have realized that.  thanks

squito commented:
I'm definitely not an expert on this, but the code seemed OK to me.  My only concern -- is it possible to add a test that ensures data is actually getting encrypting?  Otherwise the unit tests could pass just by making `encrypt = true` a no-op.  I can't think of any way to do that, though -- maybe its not possible.

vanzin commented:
/cc @rxin @aarondav in case you guys missed the JIRA update.

aarondav commented:
Will take a look tomorrow, thanks for the ping!

aarondav commented on core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala:
```diff
@@ -49,18 +49,21 @@ class NettyBlockTransferService(conf: SparkConf, securityManager: SecurityManage
-      if (!authEnabled) {
-        (nettyRpcHandler, None)
+    val rpcHandler = new NettyBlockRpcServer(serializer, blockDataManager)
+    val (serverBootstrap: Option[TransportServerBootstrap],
```
nit: This syntax gets unwieldy with more code in here, I'd advise switching to something like
```scala
var serverBootstrap: Option[TransportServerBootstrap] = None
var clientBootstrap: Option[TransportClientBootstrap] = None
if (authEnabled) {
  serverBootstrap = Some(new SaslServerBootstrap(transportConf, securityManager)
  clientBootstrap = Some(new SaslClientBootstrap(transportConf, conf.getAppId, securityManager,
    securityManager.isSaslEncryptionEnabled()))
}
```

uses the dreaded `var` but I think it may be less syntactically cumbersome.

aarondav commented on network/common/src/main/java/org/apache/spark/network/TransportContext.java:
```diff
@@ -55,14 +56,14 @@
 
   private final TransportConf conf;
-  private final RpcHandler rpcHandler;
+  private final RpcHandler appRpcHandler;
```
Why was this guy renamed? What is the app?

aarondav replied:
Ah -- I see, this is the guy before the bootstraps are executed, which is used by the client directly and as the original handler otherwise.

aarondav commented on network/common/src/main/java/org/apache/spark/network/util/TransportConf.java:
```diff
@@ -105,4 +105,19 @@ public boolean lazyFileDescriptor() {
+   * Maximum number of bytes to be encrypted at a time when SASL encryption is enabled.
+   */
+  public int maxSaslEncryptedBlockSize() {
+    return conf.getInt("spark.network.sasl.max_encrypted_block_size_kb", 64) * 1024;
```
nit: spark config style is camelCase for things which are not separated by periods

aarondav commented:
This is looking very good to me. I've reviewed the core transport part and I like the API. I will have to defer reviewing the rest of the SASL Encryption side and the tests for a bit longer, though, so apologies on that.

Commit:
Review feedback.

Commit:
Add a test that makes sure encryption is actually enabled.

SparkQA commented:
[Test build #30426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30426/consoleFull) for   PR 5377 at commit [`7fe1489`](https://github.com/apache/spark/commit/7fe148953b09e8f7c3fad66774ce6c8aa1850a45).

SparkQA commented:
[Test build #30426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30426/consoleFull) for   PR 5377 at commit [`7fe1489`](https://github.com/apache/spark/commit/7fe148953b09e8f7c3fad66774ce6c8aa1850a45).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30426/
Test PASSed.

Commit:
Fix test, SASL backend disposal.

The assert triggered by EncryptionCheckerBootstrap was being swallowed
by the netty backend, so we need to go back to checking it inline with
the test code.

Also, while investigating that, I noticed that many times the channel
handler's close() method was not called on the server side, which would
lead to the SASL backend leaking. So move the cleanup code to the
handleRemoved() method, which seems to always be called.

vanzin commented:
A note about the last commit: I noticed while debugging that calling `TransportServer.close()` doesn't close client connections that are still open. That may or may not be a bug in practice, but it could lead to malicious clients hogging resources in case the server process stays alive after the server socket is closed.

That is something for a different change, though. One way to fix it could be to have a timeout on the SASL negotiation: if the negotiation is successful we trust the client, otherwise the timeout closes the door on malicious clients.

SparkQA commented:
[Test build #30434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30434/consoleFull) for   PR 5377 at commit [`9908ada`](https://github.com/apache/spark/commit/9908ada3e2e07f4e725c76174d7b7d2e0f6a7a23).

SparkQA commented:
[Test build #30434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30434/consoleFull) for   PR 5377 at commit [`9908ada`](https://github.com/apache/spark/commit/9908ada3e2e07f4e725c76174d7b7d2e0f6a7a23).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30434/
Test PASSed.

Commit:
Remove unused import.

SparkQA commented:
[Test build #30497 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30497/consoleFull) for   PR 5377 at commit [`4797519`](https://github.com/apache/spark/commit/4797519b16c237c2573b235b49b656650eb9d0ae).

SparkQA commented:
[Test build #30497 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30497/consoleFull) for   PR 5377 at commit [`4797519`](https://github.com/apache/spark/commit/4797519b16c237c2573b235b49b656650eb9d0ae).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30497/
Test PASSed.

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java:
```diff
@@ -0,0 +1,260 @@
+      int length = msg.readableBytes();
+      if (msg.hasArray()) {
+        data = msg.array();
+        offset = msg.arrayOffset();
```
should we advance msg by the readableBytes?

vanzin replied:
It's unnecessary since `MessageToMessageDecoder` will release the input message when this method returns.

aarondav replied:
I see, it's just slightly odd that only one of the two cases moves msg's reader index. In the current implementation, you're right that it looks correct, but a seemingly unrelated change could cause issues down the line due to the discrepancy, and only in a very weird circumstance (off-heap vs on-heap messages). Consider using `msg.getBytes(msg.readerIndx(), data)` below.

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java:
```diff
@@ -0,0 +1,260 @@
+    private void nextChunk() throws IOException {
+      byteChannel.reset();
+      if (isByteBuf) {
+        int copied = byteChannel.write(buf.nioBuffer());
```
Could this be a ByteArrayOutputStream? If it were, I think you could use buf.readBytes(byteStream) or something of that sort.

vanzin replied:
Hmm... I could, but then I'd need different logic for the `FileRegion` case (since it needs a `WritableByteChannel`).

It's not a lot of code, but is more than just leaving the `skipBytes` call there.

(I could also implement `GatheringByteChannel` instead of `WritableByteChannel`, but, again, that would be more code than just the one skip call...)

aarondav replied:
I think `Channels.newChannel(outputStream)` might work.

(The reason I'd especially like this solution is that we could keep ByteArrayWritableChannel in src/test if we could just resolve this use-case without it.)

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java:
```diff
@@ -0,0 +1,260 @@
+
+        target.write(currentChunk);
+        if (!currentChunk.hasRemaining()) {
+          // Only update the count of written bytes once a full chunk has been written.
```
This mechanism may add even more performance burden, as I believe there is special logic which checks for zero-sized writes and backs off from trying to send until the channel appears writable again. However, I can't think of a good way to avoid this that isn't significantly hackier, so it's probably fine.

vanzin replied:
I see. I could return `1` instead, and keep the counts synchronized, but some pathological case might still require returning `0`. That should be rare, though, so let me do this to avoid the performance hit in the common case.

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslRpcHandler.java:
```diff
@@ -46,19 +47,30 @@
+      Channel channel,
+      RpcHandler delegate,
+      SecretKeyHolder secretKeyHolder) {
+    this.conf = conf;
```
nit: reorder fields to follow same order as constructor parameters

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java:
```diff
@@ -60,13 +60,19 @@
    */
   static final Map<String, String> SASL_PROPS = ImmutableMap.<String, String>builder()
-    .put(Sasl.QOP, "auth")
     .put(Sasl.SERVER_AUTH, "true")
```
Is this property relevant for the client? Potentially we could just do away with this static map if not.

vanzin replied:
I don't think it applies to the client. I'm also not sure whether it's needed at all, but I'll change the code so it's only set for the server.

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java:
```diff
@@ -60,13 +60,19 @@
-   * authentication only, we are not supporting integrity or privacy protection of the
-   * communication channel after authentication. This could be changed to be configurable
-   * in the future.
+   * QOP value that includes encryption.
```
I liked the spelled-out "quality of protection" better :)

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java:
```diff
@@ -75,11 +81,20 @@
     this.secretKeyId = secretKeyId;
     this.secretKeyHolder = secretKeyHolder;
+
+    String qop = alwaysEncrypt ? QOP_AUTH_CONF : String.format("%s,%s", QOP_AUTH_CONF, QOP_AUTH);
```
I assume this is a comma-separated list of the supported formats, for negotiation? Maybe add a comment to this effect.

aarondav commented on network/common/src/test/resources/log4j.properties:
```diff
@@ -23,5 +23,5 @@ log4j.appender.file.file=target/unit-tests.log
 log4j.appender.file.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{1}: %m%n
 
-# Silence verbose logs from 3rd-party libraries.
+# Filter debug messages from noisy 3rd-party libs.
```
kind of a funny change

aarondav commented on network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java:
```diff
@@ -58,10 +60,15 @@
+      boolean saslEnabled,
+      boolean saslEncryptionEnabled) {
+    Preconditions.checkArgument(
+        !saslEncryptionEnabled || saslEnabled,
```
nit: I think 2 space indent is appropriate here

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java:
```diff
@@ -0,0 +1,260 @@
+import org.apache.spark.network.util.ByteArrayWritableChannel;
+import org.apache.spark.network.util.NettyUtils;
+
+class SaslEncryption {
```
Please add a class comment

aarondav commented on network/common/src/main/java/org/apache/spark/network/TransportContext.java:
```diff
@@ -99,9 +108,9 @@ public TransportServer createServer() {
    * ChannelHandler to ensure all users of the same channel get the same TransportClient object.
    */
-  public TransportChannelHandler initializePipeline(SocketChannel channel) {
+  public TransportChannelHandler initializePipeline(SocketChannel channel, RpcHandler rpcHandler) {
```
The rpcHandler vs appRpcHandler thing is definitely confusing to someone reading this for the first time, please add a comment here or in TransportContext about the difference.

aarondav commented on network/common/src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java:
```diff
@@ -86,4 +117,237 @@ public void testNonMatching() {
+  }
+
+  @Test
+  public void testSaslEncryption() throws Exception {
```
I think these methods should be right next to testBasicSasl's definition

aarondav commented:
LGTM, only minor comments. The tests look good. Apologies for taking so long to review!

Commit:
Review feedback.

Commit:
Avoid trip through channel selector by not reporting 0 bytes written.

Commit:
Avoid exposing ByteArrayWritableChannel outside of test code.

Commit:
Skip bytes in decode path too.

vanzin commented:
I made the changes to report non-zero written bytes (https://github.com/vanzin/spark/commit/346f8291a91b0b73742263d7a23ee4027d76adb9). But after writing and testing it, I kinda started questioning whether it will improve anything...

`0` or `1` are basically returned in the same situation: the underlying channel does not have enough buffer. So returning non-zero would cause netty to "busy loop" and call `transferTo` again, which would fail to write anything and thus would now return 0.

But I guess this makes the code behave more like it would with other non-hacky `FileRegion` implementations.

SparkQA commented:
[Test build #31073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31073/consoleFull) for   PR 5377 at commit [`73cff0e`](https://github.com/apache/spark/commit/73cff0ecb6629f2308c11eff2644423bd528acda).

vanzin commented:
> Can probably eliminate this by constructing it in nextChunk(), I believe that's the only place it's used

That would generate more garbage (a new instance per call to `nextChunk`). Probably not a big deal, though...

Commit:
Minor cleanups.

SparkQA commented:
[Test build #31075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31075/consoleFull) for   PR 5377 at commit [`3d1695d`](https://github.com/apache/spark/commit/3d1695dbcf254b181886f67162882ee0ce88b6c7).

Commit:
Add test to verify chunking.

Commit:
Revert "Avoid exposing ByteArrayWritableChannel outside of test code."

This reverts commit 318ad231600884725413b5dd83c56ff73821fdf1.

Conflicts:
	network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java

aarondav commented on network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java:
```diff
@@ -0,0 +1,302 @@
+     */
+    @Override
+    public long count() {
+      return isByteBuf ? buf.readableBytes() : region.count();
```
won't this count() change as we read bytes?

vanzin replied:
It's unclear from the docs, but in the existing implementations (e.g. LazyFileRegion), `count()` always returns the same value. Also, netty does [this](https://github.com/netty/netty/blob/33f75d374091946058fc334c3cdbcd0f0a59d9b3/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L189):

    boolean done = region.transfered() >= region.count();

Which implies that `count()` should not change as bytes are transferred.

SparkQA commented:
[Test build #31073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31073/consoleFull) for   PR 5377 at commit [`73cff0e`](https://github.com/apache/spark/commit/73cff0ecb6629f2308c11eff2644423bd528acda).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `

 * This patch does not change any dependencies.

Commit:
Add comment.

SparkQA commented:
[Test build #31084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31084/consoleFull) for   PR 5377 at commit [`2f92237`](https://github.com/apache/spark/commit/2f9223729d0fedb9d16015bbc885867af6e742da).

Commit:
Clean up some unneeded changes.

SparkQA commented:
[Test build #31089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31089/consoleFull) for   PR 5377 at commit [`7a2a805`](https://github.com/apache/spark/commit/7a2a805b35f87040e8c5ea9b82d50ef6cf965d95).

SparkQA commented:
[Test build #31075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31075/consoleFull) for   PR 5377 at commit [`3d1695d`](https://github.com/apache/spark/commit/3d1695dbcf254b181886f67162882ee0ce88b6c7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `

 * This patch does not change any dependencies.

SparkQA commented:
[Test build #31089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31089/consoleFull) for   PR 5377 at commit [`7a2a805`](https://github.com/apache/spark/commit/7a2a805b35f87040e8c5ea9b82d50ef6cf965d95).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch **adds the following new dependencies:**
   * `commons-math3-3.4.1.jar`
   * `snappy-java-1.1.1.7.jar`
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`

 * This patch **removes the following dependencies:**
   * `commons-math3-3.1.1.jar`
   * `snappy-java-1.1.1.6.jar`
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`

SparkQA commented:
[Test build #31084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31084/consoleFull) for   PR 5377 at commit [`2f92237`](https://github.com/apache/spark/commit/2f9223729d0fedb9d16015bbc885867af6e742da).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

rxin commented:
@vanzin Give me a day or two. I'm checking with Netty guys on the overall design pattern. I'm worried about CPU heavy encryption blocking the IO thread calling transferTo. One solution might be to merge this, and follow up with a big TODO and JIRA to improve it.

Commit:
Merge branch 'master' into SPARK-6229

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31165/consoleFull) for   PR 5377 at commit [`47d4aff`](https://github.com/apache/spark/commit/47d4aff51538e09252b0f6286dd0e49052b07c18).

SparkQA commented:
[Test build #31165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31165/consoleFull) for   PR 5377 at commit [`47d4aff`](https://github.com/apache/spark/commit/47d4aff51538e09252b0f6286dd0e49052b07c18).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31165/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-6229

Commit:
Use fancy new size config style.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31299/consoleFull) for   PR 5377 at commit [`ff01966`](https://github.com/apache/spark/commit/ff019662055edeaf8d709087e0279afa37011353).

SparkQA commented:
[Test build #31299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31299/consoleFull) for   PR 5377 at commit [`ff01966`](https://github.com/apache/spark/commit/ff019662055edeaf8d709087e0279afa37011353).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SaslEncryption `
  * `  static class EncryptedMessage extends AbstractReferenceCounted implements FileRegion `
  * `class SaslRpcHandler extends RpcHandler `
  * `public class SaslServerBootstrap implements TransportServerBootstrap `
  * `public class SparkSaslClient implements SaslEncryptionBackend `
  * `public class SparkSaslServer implements SaslEncryptionBackend `
  * `public class ByteArrayWritableChannel implements WritableByteChannel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31299/
Test PASSed.

vanzin commented:
>  I'm worried about CPU heavy encryption blocking the IO thread calling transferTo

Yeah, that's the part that sucks with the current approach. But I couldn't find an alternative.

- ThreadPerChannelEventLoop can solve the problem but doesn't sound like the right approach.
- Using a separate thread pool for the encryption / decryption handlers is possible, but we still need to guarantee ordering when pushing messages on the outbound path; I don't think you're allowed to call blocking methods (like `ChannelPromise.sync()`) in channel threads.

Alternative suggestions are mostly welcome, if available.

rxin commented:
Thanks. I'm merging this in master.

JoshRosen commented:
How do we enable this feature? Would be good to add some docs for this.
